### DISCUSSION TITLE
What does `rawValues` in the resolver option do? (e.g. Zod)

### DISCUSSION THREAD

**Discussion by bennettdams** [react-hook-form/react-hook-form#8897] [Q&A] [answered]
Hey, quick question regarding the `rawValues` option of the Zod resolver: What does it actually do? 😅

I saw it being used in the [tRPC Kitchen Sink example](https://github.com/trpc/examples-kitchen-sink/blob/main/src/feature/react-hook-form/index.tsx#L21), but I have found any explanation for it in the [resolvers README](https://github.com/react-hook-form/resolvers), and also not in RHF docs:

![image](https://user-images.githubusercontent.com/29319414/185960201-88f34113-54d7-407f-a0fe-f255e1f2d3ad.png)

There's the following comment in the [Kitchen Sink example](https://github.com/trpc/examples-kitchen-sink/blob/main/src/feature/react-hook-form/index.tsx#L21):

_"This makes it so we can use `.transform()`s on the schema without same transform getting applied again when it reaches the server"._

There's also a description from the type definition:

![image](https://user-images.githubusercontent.com/29319414/185960686-22f18420-84f1-4fa4-811c-2ca11f74a4a1.png)

Ok, understood both, but how is this working? 

And why is this needed? I guess because you maybe want to do additional stuff in the UI, that shouldn't be done on the server when validating? Are there caveats using this property?

 ...and couldn't (or shouldn't) one just clone the schema for the UI then, instead of the ominous `rawValues` config?

There is only [one other discussion I found for this topic](https://github.com/react-hook-form/react-hook-form/discussions/8797): It uses `rawValues`, but I'm not sure why.

**Comment by leapful** [comment, marked as answer]
@bennettdams : I think you are mis-understanding documentation for Resolver of react-hook-form (from https://react-hook-form.com/ts#Resolver) and the documentation of utility library @hookform/resolvers (from https://github.com/react-hook-form/resolvers).

The library "@hookform/resolvers" is a set of resolver prepared for react-hook-form so you don't have to re-implement a resolver for native validation or for any 3rd party library schema validation (like yup, zod ...).

The option "rawValues" is purely for 3rd party library schema validation which might follow "parse then transform" paradigm where the original object data is parsed and transformed to only data related to the defined schema.

For example:

```
schema = {
  firstName: string;
  lastName: string;
}

originalObject = {
  firstName: 'A',
  lastName: 'B',
  age: 25
}

const parsed = zod.schema.parse(originalObject)

// where "parsed" is:

{
  firstName: 'A',
  lastName: 'B'
}
```

The purpose is simple, you don't want your original data (some fields might be not related to form data) is altered after validating .

**Reply by bennettdams** [reply]
Hey @leapful,
thanks for the answer! I'm well aware of what the resolvers are and what they're used for. It's just that I think an option that exists (like `rawValues`) for the resolver should be at least described _somewhere_ and as resolvers are also part of the React Hook Form docs, I mentioned them together with the "resolvers README" above.

With your explanation for `rawValues` now the [comment from the kitchen sink example](https://github.com/trpc/examples-kitchen-sink/blob/main/src/feature/react-hook-form/index.tsx#L20) also makes sense to me. It basically means that `rawValues` makes sure we use a different schema on the client and the server, because when using `rawValues`, the (tRPC) server will receive the raw values, instead of the transformed ones.

So a better example is e.g. having a date field. Maybe it's a `string` initially, but you use `transform` to make it a real `Date` object, as you maybe need it like that for the UI.
When you submit that form, you don't want to send a `Date` object to the server, but rather the initial `string` value, as that is how the schema expects this field to be. Using `rawValues` gives you the `string` value (= value without `transform`) when submitting.

Again, maybe this should be described in (any of 😅) the docs?
